### PR TITLE
add enclose_binary function

### LINKED ISSUE #38: Recursive range evaluation
cc: @dfcaporale

### PR DESCRIPTION
**Pull request by dfcaporale**
Closes #38

### CODE REVIEW

File: src/branchandbound.jl [hunk 1]
```diff
@@ -68,3 +68,25 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
     R = append!(RR, R[(index + 1):length(R)])
     return D, R
 end
+
+function enclose_binary(f, dom; kmax=3, tol=1e-3)
```
Comment (mforets) [inline src/branchandbound.jl]: I think that this function assumes that the domain is an inteval (1D). In that case we should rather add the type annotation `dom::Interval`.

Comment (mforets) [reply src/branchandbound.jl]: Let's also add an optional argument `algorithm` that receives the algorithm (with the default being `:IntervalArithmetic`).

File: src/branchandbound.jl [hunk 2]
```diff
@@ -68,3 +68,25 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
+    y = f(dom)
```
Comment (mforets) [inline src/branchandbound.jl]: The algorithm can be used here, changing `y = f(dom)` with `y = enclose(f, dom, algorithm=algorithm)`.

File: src/branchandbound.jl [hunk 3]
```diff
@@ -68,3 +68,25 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
+    yinf, ysup = inf(y), sup(y)
+    x = bisect(dom)
+    ynew = union(f(x[1]), f(x[2]))
+    ynew_inf, ynew_sup = inf(ynew), sup(ynew)
+    no_cambia_inf = abs(yinf - ynew_inf) < tol
```
Comment (mforets) [inline src/branchandbound.jl]: Can you change these variable names to english?

File: src/branchandbound.jl (hunk 1 above)
Comment (dfcaporale) [inline src/branchandbound.jl]:
```suggestion
function enclose_binary(f, dom::Interval; kmax=3, tol=1e-3)
```

File: src/branchandbound.jl [hunk 4]
```diff
@@ -68,3 +68,25 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
     R = append!(RR, R[(index + 1):length(R)])
     return D, R
 end
+
+function enclose_binary(f, dom::Interval; kmax=3, tol=1e-3, algorithm=:IntervalArithmetic)
+    y = enclose(f, dom, algorithm=algorithm)
+    yinf, ysup = inf(y), sup(y)
+    x = bisect(dom)
+    ynew = union(f(x[1]), f(x[2]))
```
Comment (mforets) [inline src/branchandbound.jl]:
```suggestion
    ynew = union(enclose(f, x[1], algorithm=algorithm), enclose(f, x[2], algorithm=algorithm))
```

File: src/branchandbound.jl [hunk 5]
```diff
@@ -68,3 +68,27 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
+    ynew = union(fx1, fx2)
+    ynew_inf, ynew_sup = inf(ynew), sup(ynew)
+    no_cambia_inf = abs(yinf - ynew_inf) < tol
+    no_cambia_sup = abs(ysup - ynew_sup) < tol
+    no_cambia_ninguno = (no_cambia_inf && no_cambia_sup)
+    empeora_inf = ynew_inf < yinf
+    empeora_sup = ynew_sup > ysup
+    empeoran_ambos = empeora_inf && empeora_sup
+    if kmax == 0 || no_cambia_ninguno || empeoran_ambos
+        return Interval(yinf, ysup)
```
Comment (mforets) [inline src/branchandbound.jl]: I would move the base condition for the recursion just after line 74. In cases like this we can do a short-circuit and, like so:

```julia
k == 0 && return Interval(yinf, ysup)
```

File: src/branchandbound.jl (hunk 3 above)
Comment (mforets) [reply src/branchandbound.jl]: Proposal:

```julia
inf_close = abs(yinf - ynew_inf) <= tol
sup_close = abs(ysup - ynew_sup) <= tol
both_close = inf_close && sup_close

inf_improves = ynew_inf > yinf
sup_improves = ynew_sup < ysup
both_improve = inf_improves && sup_improves

if both_close || !both_improve
    return Interval(yinf, ysup)
end
```

File: src/branchandbound.jl [hunk 6]
```diff
@@ -68,3 +68,28 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
+    inf_close = abs(yinf - ynew_inf) <= tol
+    sup_close = abs(ysup - ynew_sup) <= tol
+    both_close = inf_close && sup_close
+    inf_improves = ynew_inf > yinf
+    sup_improves = ynew_sup < ysup
+    both_improve = inf_improves && sup_improves
+    if both_close || !both_improve
+        return Interval(yinf, ysup)
+    end
+    yinf = max(yinf, ynew_inf)
```
Comment (mforets) [inline src/branchandbound.jl]: Here we don't need to evaluate the `max` and `min` again, so we can do:

```julia
if inf_improves
    yinf = ynew_inf
end

if sup_improves
    ysup = ynew_sup
end
```

File: src/branchandbound.jl [hunk 7]
```diff
@@ -68,3 +68,34 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
+    ysup = min(ysup, ynew_sup)
+    if inf_improves
+        yinf = ynew_inf
+    end
+    if sup_improves
+        ysup = ynew_sup
+    end
+    e1 = enclose_binary(f, x[1], kmax=kmax-1, algorithm=algorithm)
+    e2 = enclose_binary(f, x[2], kmax=kmax-1, algorithm=algorithm)
+    return Interval(min(inf(e1), inf(e2)), max(sup(e1), sup(e2)))
```
Comment (mforets) [inline src/branchandbound.jl]:
```suggestion
    return Interval(hull(e1, e2))
```

File: src/branchandbound.jl [hunk 8]
```diff
@@ -68,3 +68,34 @@ function divide_dom!(p::Union{TaylorN{T}, Taylor1{T}},
...
 end
+
+function enclose_binary(f, dom::Interval; kmax=3, tol=1e-3, algorithm=:IntervalArithmetic)
+    y = enclose(f, dom, algorithm=algorithm)
+    yinf, ysup = inf(y), sup(y)
+    kmax == 0 && return Interval(yinf, ysup)
+    x = bisect(dom)
+    fx1 = enclose(f, x[1], algorithm=algorithm)
+    fx2 = enclose(f, x[2], algorithm=algorithm)
+    ynew = union(fx1, fx2)
```
Comment (mforets) [inline src/branchandbound.jl]:
```suggestion
    ynew = hull(fx1, fx2)
```

**Pull request merged**